### DISCUSSION TITLE
Using syslog facility to log to custom file

### DISCUSSION THREAD

**Discussion by ozcanay** [gabime/spdlog#2123] [Q&A] [answered]
I am trying to write to "custom_file.log" file in /var/log directory. My OS is CentOS. What I have tried to far:

Adding line 

`local2.*                                                /var/log/custom_file.log`

to "/etc/rsyslog.conf" file. Then entering command "sudo service rsyslog restart" to restart rsyslog.

After that, I am just using the most basic example for syslog facility of spdlog:

```
std::string ident = "spdlog-example";
auto syslog_logger = spdlog::syslog_logger_mt("syslog", ident, LOG_LOCAL2);
syslog_logger->warn("This is warning that will end up in syslog.");
```

Note that I am using LOG_LOCAL2 syslog facility here. However, the message won't be printed to "/var/log/custom_file.log".
If I change LOG_LOCAL2 to LOG_PID, the message will be printed to "/var/log/messages".

To see if LOG_LOCAL2 facility is not at fault, I tried logging via Linux syscalls:

```
#define open_log(str) do {                                                             \
    openlog(str, LOG_CONS | LOG_NDELAY, LOG_LOCAL2);                                   \
} while(0)

#define info_log(str, ...) do {                                                        \
    syslog(LOG_INFO, str,  ##__VA_ARGS__);                                             \
} while(0)

int main()
{
    open_log("config");

    info_log("TEST1");
    info_log("TEST2");
    info_log("TEST3");
}
```

This works as intended.

So, my question is, first of all, is it possible to write to a custom log file via spdlog. If so, where exactly am I wrong here?

**Comment by tt4g** [comment, marked as answer]
First of all, I don't usually use syslog, so I may be guiding you wrongly.

Looking at the function declaration, it is the fourth argument of `spdlog::syslog_logger_mt()` that corresponds to the third argument `facility` of `openlog()`.


https://github.com/gabime/spdlog/blob/eb3220622e73a4889eee355ffa37972b3cac3df5/include/spdlog/sinks/syslog_sink.h#L95-L101

https://github.com/gabime/spdlog/blob/eb3220622e73a4889eee355ffa37972b3cac3df5/include/spdlog/sinks/syslog_sink.h#L24-L37

`A` should be specified as the fourth argument.

```diff
std::string ident = "spdlog-example";
auto syslog_logger = spdlog::syslog_logger_mt("syslog", ident, LOG_CONS | LOG_NDELAY, LOG_LOCAL2);
syslog_logger->warn("This is warning that will end up in syslog.");
```

**Reply by ozcanay** [reply]
Yeah, that was my bad. Thanks for the help!

`spdlog::syslog_logger_mt("syslog", "config-manager", LOG_CONS | LOG_NDELAY, LOG_LOCAL2);`

works well.
